tensorflow/core/kernels/data/iterator_ops.cc

```diff
@@ -581,7 +584,20 @@ class ToSingleElementOp : public HybridAsyncOpKernel {
       return errors::InvalidArgument("Dataset was empty.");
     }
     for (int i = 0; i < components.size(); ++i) {
-      // TODO(mrry): Check that the shapes match the shape attrs.
+      if (components[i].dtype() != output_types_[i]) {
+        return errors::InvalidArgument(
+            "The result does not match the expected type for "
+            "component ",
+            i, ". Expected: ", DataTypeString(output_types_[i]),
+            ". Actual: ", DataTypeString(components[i].dtype()), ".");
+      }
+      if (!output_shapes_[i].IsCompatibleWith(components[i].shape())) {
+        return errors::InvalidArgument(
+            "The result does not match the expected shape "
+            "for component ",
+            i, ". Expected: ", output_shapes_[i].DebugString(),
+            ". Actual: ", components[i].shape().DebugString(), ".");
+      }
```

Comment:
Use `VerifyTypesMatch` and `VerifyShapesCompatible` instead.

Reply:
> Use `VerifyTypesMatch` and `VerifyShapesCompatible` instead.

@jsimsa Using these two functions need to allocate vectors to save the output_shapes and output_types for `components`. I'm afraid they will add overhead to those kernels.

Reply:
You could achieve what I am asking for (which in essence is to avoid code duplication) by the following refactoring:

```
Status VerifyTypesMatch(const DataType& expected, const DataType& received, int index) {
  if (expected != received) {
    return errors::InvalidArgument("Data type mismatch at component ", index,
        ": expected ", DataTypeString(expected), " but got ", DataTypeString(received), ".");
  }
  return Status::OK();
}

Status VerifyTypesMatch(const DataTypeVector& expected, const DataTypeVector& received) {
  if (expected.size() != received.size()) {
    return errors::InvalidArgument(
        "Number of components does not match: expected ", expected.size(),
        " types but got ", received.size(), ".");
  }
  for (size_t i = 0; i < expected.size(); ++i) {
    TF_RETURN_IF_ERROR(VerifyTypesMatch(expected[i], received[i], i));
  }
  return Status::OK();
}

Status VerifyTypesMatch(const DataTypeVector& expected, const std::vector<Tensor>& received) {
  if (expected.size() != received.size()) {
    return errors::InvalidArgument(
        "Number of components does not match: expected ", expected.size(),
        " types but got ", received.size(), ".");
  }
  for (size_t i = 0; i < expected.size(); ++i) {
    TF_RETURN_IF_ERROR(VerifyTypesMatch(expected[i], received[i].dtype(), i));
  }
  return Status::OK();
}
```

(and similarly for `VerifyShapesCompatible`). 

Once you perform this refactoring you will be able to avoid code duplication (without incurring allocation). Feel free to also update `ReduceDatasetOp::DoCompute` to replace the currently duplicating code for type and shape checking with the new utilities.

Reply:
Refactor made.